src/tests/end2end/DepthStencilStateTests.cpp

```diff
@@ -0,0 +1,512 @@
+                .SetExtent(kRTSize, kRTSize, 1)
+                .SetFormat(nxt::TextureFormat::D32FloatS8Uint)
+                .SetMipLevels(1)
+                .SetAllowedUsage(nxt::TextureUsageBit::OutputAttachment)
+                .SetInitialUsage(nxt::TextureUsageBit::OutputAttachment)
+                .GetResult();
+
+            depthTextureView = depthTexture.CreateTextureViewBuilder().GetResult();
+
+            framebuffer = device.CreateFramebufferBuilder()
+                .SetRenderPass(renderpass)
+                .SetAttachment(0, renderTargetView)
+                .SetAttachment(1, depthTextureView)
+                .SetDimensions(kRTSize, kRTSize)
+                .GetResult();
+
+            vsModule = utils::CreateShaderModule(device, nxt::ShaderStage::Vertex, R"(
+                #version 450
+                layout(set = 0, binding = 0) uniform myBlock {
+                    vec3 color;
+                    float depth;
+                } myUbo;
+                void main() {
+                    const vec2 pos[6] = vec2[6](
+                        vec2(-1.f, 1.f), vec2(-1.f, -1.f), vec2(1.f, -1.f), // front-facing
+                        vec2(-1.f, 1.f), vec2(1.f, 1.f), vec2(1.f, -1.f)    // back-facing
+                    );
+                    gl_Position = vec4(pos[gl_VertexIndex], myUbo.depth, 1.f);
+                }
+            )");
+
+            fsModule = utils::CreateShaderModule(device, nxt::ShaderStage::Fragment, R"(
+                #version 450
+                layout(set = 0, binding = 0) uniform myBlock {
+                    vec3 color;
+                    float depth;
+                } myUbo;
+                layout(location = 0) out vec4 fragColor;
+                void main() {
+                    fragColor = vec4(myUbo.color, 1.f);
+                }
+            )");
+
+            bindGroupLayout = device.CreateBindGroupLayoutBuilder()
+                .SetBindingsType(nxt::ShaderStageBit::Vertex | nxt::ShaderStageBit::Fragment, nxt::BindingType::UniformBuffer, 0, 1)
+                .GetResult();
+
+            pipelineLayout = device.CreatePipelineLayoutBuilder()
+                .SetBindGroupLayout(0, bindGroupLayout)
+                .GetResult();
+        }
+
+        struct TestSpec {
+            const nxt::DepthStencilState& depthStencilState;
+            RGBA8 color;
+            float depth;
+            uint32_t stencil;
+        };
+
+        // Draw a base triangle at z-depth 0.5
+        // Execute three tests: one draws a triangle at depth 0, the second test at depth 0.5, and the third at depth 1.0
+        // (the base triangle is always drawn first)
+        // Apply the given depth comparison function to the triangle
+        // The less, equal, greater booleans denote wether the respective triangle should be visible based on the comparison function
+        void CheckDepthCompareFunction(nxt::CompareFunction compareFunction, bool less, bool equal, bool greater) {
```

Comment:
nit: Above a function, it is better to explain what it does and why, the how it does it can be inside. That is, unless the how is extremely important for the caller, but this is not the case here. Same for other helper functions below.